Documentation/coding-guidelines/interop-pinvokes.md

```diff
@@ -0,0 +1,238 @@
+P/Invokes
+=========
+
+This document extends the [Interop Guidelines](interop-guidelines.md) to provide more specific guidelines, notes, and resources for defining P/Invokes.
+
+Guidelines
+----------
+1. Avoid StringBuilder
+2. Use `ExactSpelling` where possible
+3. Only use `SetLastError` where appropriate
+4. Be careful of `BOOL` vs `BOOLEAN`
+5. Watch for buffer size for output strings (null/no null)
+6. Do not create your own buffer cache, use `ArrayPool`
+7. Use blittable types in struct definitions where possible
+
+Attributes
+----------
+
+**Implicit attributes applied to parameter and return values:**
+
+|                  | Implicit Attribute |
+|------------------|--------------------|
+| parameter        | `[In]`             |
+| `out` parameter  | `[Out]`            |
+| `ref` parameter  | `[In],[Out]`       |
+| return value     | `[Out]`            |
+
+**`[DllImport()]` [1] attribute settings:**
+
+| Setting | Recommendation | Details |
+|---------|----------------|---------|
+| [`PreserveSig`][2]   | keep default       | When this is explicitly set to false (the default is true), failed HRESULT return values will be turned into Exceptions (and the return value in the definition becomes null as a result).|
+| [`SetLastError`][3]  | as per API         | Set this to true (default is false) if the API uses GetLastError and use Marshal.GetLastWin32Error to get the value. If the API sets a condition that says it has an error, get the error before making other calls to avoid inadvertently having it overwritten.|
+| [`ExactSpelling`][4] | `true`             | Set this to true (deafult is false) and gain a slight perf benefit as the framework will avoid looking for an "A" or "W" version. (See NDirectMethodDesc::FindEntryPoint).|
```

Comment:
It does look like `FindEntryPoint` does that searching regardless of OS, though.